pkg/server/config/config.go

```diff
@@ -121,3 +131,19 @@ func BuildTLS(opts TLS) (*tls.Config, error) {
 
 	return config, nil
 }
+
+func (ap *AuthenticatingProxy) fixSpiffeIds() []int {
```

Comment:
I'm not sure I fully understand the config load behavior. If there's a bad SPIFFE ID in yaml file, does it get omitted and an empty struct is present in the array instead? If so, why do we attempt to trim the list of SPIFFE IDs if `Load` is going to return an error any way? Maybe the answer here is to simply turn this function into `validateSPIFFEIDs` that returns an error (and provides indices of SPIFFE IDs that caused the error, etc.)?

Reply:
I think the attempt to trim was an overabundance of caution in the case that an error doesn't result in the program failing. If the program doesn't fail for some reason or another (downstream changes maybe?), we'd see certs being accepted with empty SPIFFE IDs. I had tried an "omitempty" flag for the YAML parser but I wasn't able to find success with that.

Unfortunately parsing the errors out cleanly here is difficult for me probably due to a lack of experience in Go and its error handling. From what I could gather, there's not really a way to discern if an error is from the SPIFFE ID parser or a more broad YAML issue completely unrelated outside of doing string comparison. Seems like SPIFFE will just throw up an errorString and YAML only has TypeError that is exported which does not match at least one case that I could find. I am personally unsure of how to really nail that bit down in a proper validation format.

Reply:
it seems reasonable to fail during `Load` if some config SPIFFE ID could not be parsed correctly. from my understanding, that's what this function is really validating and it might just be the name that's throwing me off

in terms of distinguishing between a yaml and spiffe error, it seems like we are trying to catch the spiffe parsing errors here, and the yaml errors in `Load` when we try to `unmarshall` the config. is that the type of errors we're talking about or am I missing something else?

Reply:
In my head I was hoping that there'd be recovery space, like in the case you've misconfigured an element we could report it and move on gracefully. I think I abandoned that thought due to the challenges of making sure it was correct.

Yeah, that's right. There's quite a few [spiffe errors](https://github.com/spiffe/go-spiffe/blob/main/v2/spiffeid/errors.go) that can be caught that they don't have exported. Most of the resources I found were relying on exported types to do some sort of meaningful handling, but in the case where those types don't exist it seems like the only option is to do string comparison which had been noted as "not change tolerant". I was trying to find the best way to avoid calling a SPIFFE ID error a YAML error or vice versa but ended up with just reporting the error and watching it fail fatally.

Reply:
Which I guess leads to the following question in my mind: is it too troublesome to have the SPIFFE ID parsed right at the YAML unmarshal or is it better to grab it as a string and then do any parsing and handling afterwards?

Reply:
I might be missing something and I'm not super familiar with handling spiffe errors, are there resources that can handle the errors if they are caught and allow the program to continue running? If it doesn't actually fix the errors, it would make sense to log it and fail fatally, especially since these are config errors so it would fail on start up when there are new deploys.

I do think it makes sense to grab it as a string and then validate it afterwards. Maybe we can add other field to `AuthenticatingProxy`, like `ValidatedSpiffeIds` with the `[]spiffeid.ID` type, and populate it after validation.

Reply:
Yeah, if there were errors exported that could help cleanly delineate between YAML errors unrelated to SPIFFE errors there could probably be some graceful cleanup done but unfortunately it doesn't seem like there's a stable way to do that so the newest commit fails fatally.

I think that validation pattern is ultimately where I think it'll end up. I think that's probably going to be a near-future task along with the other parameters that could use validation.